src/tools/rbd_mirror/image_map/Types.h

```diff
@@ -0,0 +1,91 @@
+// -*- mode:C++; tab-width:8; c-basic-offset:2; indent-tabs-mode:t -*-
+// vim: ts=8 sw=2 smarttab
+
+#ifndef CEPH_RBD_MIRROR_IMAGE_MAP_TYPES_H
+#define CEPH_RBD_MIRROR_IMAGE_MAP_TYPES_H
+
+#include <map>
+#include <string>
+#include <boost/variant.hpp>
+
+#include "include/buffer.h"
+#include "include/encoding.h"
+#include "tools/rbd_mirror/types.h"
+
+namespace ceph {
+class Formatter;
+}
+
+namespace rbd {
+namespace mirror {
+namespace image_map {
+
+enum MirrorPolicyMetaType {
+  MIRROR_POLICY_META_TYPE_NONE = 0,
+};
+
+struct MirrorPolicyMetaTypeNone {
+  static const MirrorPolicyMetaType TYPE = MIRROR_POLICY_META_TYPE_NONE;
+
+  MirrorPolicyMetaTypeNone() {
+  }
+
+  void encode(bufferlist& bl) const {
+  }
+
+  void decode(__u8 version, bufferlist::iterator& it) {
+  }
+
+  void dump(Formatter *f) const {
+  }
+};
+
+struct MirrorPolicyMetaTypeUnknown {
+  static const MirrorPolicyMetaType TYPE = static_cast<MirrorPolicyMetaType>(-1);
+
+  MirrorPolicyMetaTypeUnknown() {
+  }
+
+  void encode(bufferlist& bl) const {
+    assert(false);
+  }
+
+  void decode(__u8 version, bufferlist::iterator& it) {
+  }
+
+  void dump(Formatter *f) const {
+  }
+};
+
+typedef boost::variant<MirrorPolicyMetaTypeNone,
+                       MirrorPolicyMetaTypeUnknown> MirrorPolicyMeta;
+
+struct MirrorPolicyData {
+  MirrorPolicyData()
+    : policy_meta(MirrorPolicyMetaTypeNone()) {
```

Comment:
why `Unknown`? wouldn't we want to have policy data empty now and later introduce tracking variables to decide what and where to migrate? Or do you mean to get rid to `None` policy meta type and just have `Unknown` for now (which as with this version of the pr has an `assert` placed deliberately in its `encode`)

Reply:
I would like to force the internal code users to explicitly pick what they are choosing to store instead of allowing it to default to a valid meta type. If the simply policy doesn't override it w/ the none meta, that should be considered a bug.